### ISSUE TITLE
PGNs 128776,128777, 128778 (Windlass)

### ISSUE THREAD

**Issue by JohnStov** [SignalK/n2k-signalk#282]
Add support for windlass PGNs

{"prio":2,"pgn":128776,"dst":255,"src":36,"timestamp":"2011-11-24-22:42:04.440","fields":{"SID":170,"Windlass ID":0,"Windlass Direction Control":0,"Speed Control":0,"Speed Control Type":0,"Anchor Docking Control":0,"Power Enable":0,"Mechanical Lock":0,"Deck and Anchor Wash":0,"Anchor Light":0,"Command Timeout":"00:00:00.85000","Windlass Control Events":[]}}

{"prio":2,"pgn":128777,"dst":255,"src":36,"timestamp":"2011-11-24-22:42:04.440","fields":{"SID":170,"Windlass ID":0,"Windlass Motion Status":0,"Rode Type Status":0,"Rode Counter Value":15.3,"Windlass Line Speed":0,"Anchor Docking Status":0,"Windlass Operating Events":[]}}

{"prio":2,"pgn":128778,"dst":255,"src":36,"timestamp":"2011-11-24-22:42:04.440","fields":{"SID":170,"Windlass ID":0,"Total Motor Time":"00:03:00","Controller voltage":12,"Motor current":15,"Windlass Monitoring Events":[]}}

**Comment by Hakansv** [comment]
A windlass rode counter is implemented in Dashboard @ Github master code. 
Please read this [discussion](https://github.com/OpenCPN/OpenCPN/discussions/3427#discussioncomment-11008140) 
According to your:  "pgn":128777,..........,"Rode Counter Value":15.3," " 
My assumption that the value (15.3) would represent how many meters have been ejected seems to be correct. So it should work.
Is it possible for you to test?
If you can't build from source please tell us what system you have and we can possible show a download link if you would like to test the alpha version of OCPN.

**Comment by JohnStov** [comment]
That's correct.

It's from my test data. I have working code but I haven't created the PR yet.
I'm in the process of building a homebrew N2K chain counter/windlass controller that will use Timo Lappalainen's NMEA2000 library(https://github.com/ttlappalainen/NMEA2000). I want this support so that I can import the data into SignalK.

If anyone has any real world  windlass PGN data I would be happy to use it in my test case.

**Comment by Hakansv** [comment]
> I want this support so that I can import the data into SignalK.

What I know SignalK does not yet handle these PGNs and there is no windlass paths available.  Please correct me if I'm wrong.

**Comment by JohnStov** [comment]
That is correct. I have raised a question on the SignalK Discourse Server about it. This is what I said :

"I'm in the process of implementing handlers in n2k-signalk for the windlass PGNs (128776, 128777, 128778). Does anyone have any suggestions for where in the full model I should root the signalk data? There is <vessel>.navigation.anchor, but that doesn't feel right because it's not really navigation information, and also NMEA2000 allow for multiple windlasses, so the final path would be something like <vessel>.navigation.anchor.<RegExp>.windlass.rodeCounter, which feels clunky. I could go with <vessel>.windlass.<RegExp>.rodeCounter but that fails the tree.should.be.validSignalKVesselIgnoringIdentity test. All suggestions welcome."

**Comment by Hakansv** [comment]
As soon SK have a path for 128777,..........,"Rode Counter Value" we will implement that to O Dashboard's Windlass instrument.
 
Look [here](https://github.com/SignalK/specification/issues?q=is%3Aissue+is%3Aopen+windlass+in%3Arirle%2Cbody) at SK issues
And maybe [here ](https://github.com/LukeSavage1306/Windlass-Remote-Chain-Counter-and-SignalK-Feed/blob/main/Windlass%20control%20with%20preset%20scope%20V2.0.ino)to not invent the wheel again?

The meter windlass is judged as a "navigational" information suitable for OCPN core Dashboard. See my shot.
But to implement a windlass control would not be suitable. Possibly it could be suitable for a plugin?


![bild](https://github.com/user-attachments/assets/efc740eb-4a43-469d-b1e4-fa2beb8e57eb)

**Comment by JohnStov** [comment]
Thank you very much! This is all really useful information. 

My intention is still to build a NMEA2000 windlass controller, so I will leave this issue open and submit a PR soon.  I will update my pull request to use the schema in PR #601, since this seems to represent current thinking on where windlasses fit in the schema.
